agent/consul/state/event.go

```diff
@@ -65,8 +125,73 @@ func (e *EventPublisher) Commit() {
 			e.lastIndex[event.Topic] = event.Index
 		}
 	}
+}
 
-	e.staged = nil
+func (e *EventPublisher) handleACLUpdate(tx *memdb.Txn, event stream.Event) error {
+	switch event.Topic {
+	case stream.Topic_ACLTokens:
+		token := event.GetACLToken()
+		subscribers, ok := e.tokenSubs[token.Token.SecretID]
```

Comment:
Accessor seems like a better fit here, it's useful to be able to iterate through the subscribers' tokens when a policy/role changes and a hash wouldn't let us do that

Reply:
I rolled this back after realizing this wouldn't work with legacy ACLs because those tokens don't have a separate accessor - we could use a hash but then we'd have to iterate through every token in the store instead of just the ones used by subscribers. What do you think?

Reply:
@mkeeler is that true? I think we auto-gen accessor ids for legacy ACLs on startup with 1.4 but I forget maybe we didn't do that or there is an interim where they might be empty.

Honestly It doesn't change our threat model so it's probably OK as it is, I just dislike the idea that secrets are used as keys in places they don't really need to be. In practice I _think_ Go will not even make a copy of the string contents here and it's really just another pointer to the same string we already have in memory so it's not even that, it's more that it would be easy to debug-print the map or something later without thinking through and end up risking exposing secrets because it wasn't obvious that the component you were debugging was sensitive.

tl;dr I'm _OK_ leaving this for simplicity but wish we could use accessor more for stuff like this.

 > it's useful to be able to iterate through the subscribers' tokens when a policy/role changes and a hash wouldn't let us do that

Maybe I'm missing something but surely you can iterate through subscribers tokens, hash each one and then lookup? I'm not sure sure it's warranted thought per above.

Reply:
What was the final conclusion @kyhavlov? I'm OK with leaving it for now I guess but it seems there are at least a couple of options here (hashing or dissallowing non-upgraded legacy tokens) that would be easy enough to do?

Reply:
Currently we're storing hashes of the secret IDs to track changes. I don't have a strong preference either way but it seems nice to be backwards compatible by using the secret. If we had to switch over to accessors in the future it wouldn't be a large change since this is purely internal and ephemeral (recreated on server restart), so it we officially stop supporting legacy ACLs at some point this wouldn't be a point of friction.